### PR TITLE
feat: compose modal add farcast channel

### PR DESCRIPTION
**Pull request by tomwang-io**
https://mask.atlassian.net/browse/FW-724

### CODE REVIEW

File: src/components/Compose/Channel/ChannelAction.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import { Popover } from '@headlessui/react';
+import { Trans, t } from '@lingui/macro';
+import { BugAntIcon, ChevronRightIcon } from '@heroicons/react/24/outline';
+import { ChannelPanel } from '@/components/Compose/Channel/ChannelPanel.js';
+import type { Channel } from '@/providers/types/SocialMedia.js';
+
+export type ChannelActionProps = {
```
Comment (guanbinrui) [inline src/components/Compose/Channel/ChannelAction.tsx:7]: We prefer using `interface`.

File: src/components/Compose/Channel/ChannelAction.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    isRootPost: boolean;
+    channelList: Channel[];
+    channel: Channel;
+    inputText: string;
+    setInputText: (v: string) => void;
+    updateChannel: (c: Channel) => void;
```
Comment (guanbinrui) [inline src/components/Compose/Channel/ChannelAction.tsx:13]:
```suggestion
    setInputText: (inputText: string) => void;
    updateChannel: (channel: Channel) => void;
```

File: src/components/Compose/Channel/ChannelPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import SearchIcon from '@/assets/search.svg';
+import YesIcon from '@/assets/yes.svg';
+import { Avatar } from '@/components/Avatar.js';
+import { SearchInput } from '@/components/Search/SearchInput.js';
+import { classNames } from '@/helpers/classNames.js';
+import { useIsSmall } from '@/hooks/useMediaQuery.js';
+import type { Channel } from '@/providers/types/SocialMedia.js';
+
+// export type SubChannel = Pick<Channel, 'id' | 'name' | 'imageUrl'>;
+interface FcChannelProps {
```
Comment (guanbinrui) [inline src/components/Compose/Channel/ChannelPanel.tsx:15]:
```suggestion
interface ChannelProps {
```

File: src/components/Compose/Channel/ChannelPanel.tsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import RadioDisableNoIcon from '@/assets/radio.disable-no.svg';
+import SearchIcon from '@/assets/search.svg';
+import YesIcon from '@/assets/yes.svg';
+import { Avatar } from '@/components/Avatar.js';
+import { SearchInput } from '@/components/Search/SearchInput.js';
+import { classNames } from '@/helpers/classNames.js';
+import { useIsSmall } from '@/hooks/useMediaQuery.js';
+import type { Channel } from '@/providers/types/SocialMedia.js';
+
+// export type SubChannel = Pick<Channel, 'id' | 'name' | 'imageUrl'>;
```
Comment (guanbinrui) [inline src/components/Compose/Channel/ChannelPanel.tsx:14]:
```suggestion
```

File: src/store/useComposeStore.ts [hunk 5]
```diff
@@ -461,3 +478,6 @@ const useComposeStateBase = create<ComposeState, [['zustand/immer', unknown]]>(
 );
 
 export const useComposeStateStore = createSelectors(useComposeStateBase);
+
+// @ts-ignore
+global.__useComposeStateBase = useComposeStateBase;
```
Comment (guanbinrui) [inline src/store/useComposeStore.ts:483]:
```suggestion
```

File: src/store/useComposeStore.ts [hunk 6]
```diff
@@ -51,6 +51,9 @@ export interface CompositePost {
     // parsed open graphs from url in chars
     openGraphs: OpenGraph[];
     rpPayload: RedPacketPayload | null;
+
+    // only available in farcaster
+    channel: Channel|null;
```
Comment (guanbinrui) [inline src/store/useComposeStore.ts:56]: Design this for all platforms.

File: src/services/postToFarcaster.ts [hunk 7]
```diff
@@ -70,6 +73,7 @@ export async function postToFarcaster(type: ComposeType, compositePost: Composit
             );
         },
         compose: (images) => {
+            console.log('xxx', composeDraft('Post', images));
```
Comment (guanbinrui) [inline src/services/postToFarcaster.ts:76]: No dev log.

File: src/services/postToFarcaster.ts [hunk 8]
```diff
@@ -52,8 +53,10 @@ export async function postToFarcaster(type: ComposeType, compositePost: Composit
                 (x) => x.url.toLowerCase(),
             ),
             commentOn: type === 'reply' && farcasterParentPost ? farcasterParentPost : undefined,
-            parentChannelKey: hasPayload ? env.external.NEXT_PUBLIC_REDPACKET_CHANNEL_KEY : undefined,
-            parentChannelUrl: hasPayload ? env.external.NEXT_PUBLIC_REDPACKET_CHANNEL_URL : undefined,
+            // parentChannelKey: hasPayload ? env.external.NEXT_PUBLIC_REDPACKET_CHANNEL_KEY : undefined,
+            // parentChannelUrl: hasPayload ? env.external.NEXT_PUBLIC_REDPACKET_CHANNEL_URL : undefined,
```
Comment (guanbinrui) [inline src/services/postToFarcaster.ts:57]: Please remove those envs.

```suggestion
```

File: src/providers/firefly/SocialMedia.ts [hunk 9]
```diff
@@ -69,7 +69,9 @@ class FireflySocialMedia implements Provider {
 
         return formatBriefChannelFromFirefly(data);
     }
-
+    /**
+     * @wan: cannot do pagination ?
+     */
```
Comment (guanbinrui) [inline src/providers/firefly/SocialMedia.ts:74]: Yes, only single page here.

```suggestion
```

File: src/components/Search/SearchBar.tsx [hunk 10]
```diff
@@ -64,6 +64,7 @@ const SearchBar = memo(function SearchBar(props: SearchBarProps) {
                 <form
                     className="w-full flex-1"
                     onSubmit={(ev) => {
+                        console.log('--->handle search');
```
Comment (guanbinrui) [inline src/components/Search/SearchBar.tsx:67]:
```suggestion
```

File: src/components/Compose/Channel/utils.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    const { type } = useComposeStateStore();
+    return availableSources.includes(SocialPlatform.Farcaster) && type === 'compose';
+}
+export function useSetDefaultSelectedChannel(kw: string) {
+    const showChannel = useShowChannel();
+    const { rpPayload, channel } = useCompositePost();
+    const { updateChannel } = useComposeStateStore();
+    const { value: fireflyChannel } = useAsync(async () => fetchFireflyGardenChannel(), []);
+    const preRpPayload = usePrevious(rpPayload);
+    useEffect(() => {
```
Comment (guanbinrui) [inline src/components/Compose/Channel/utils.ts:59]: Violation of the single data flow principle.

File: src/components/Compose/Channel/ChannelPanel.tsx [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+interface FcChannelProps {
+    channelList: Channel[];
+    selectChannel: (channel: Channel) => void;
+    selectedChannel: Channel | null;
+    inputText: string;
+    setInputText: (v: string) => void;
```
Comment (nuanyang233) [inline src/components/Compose/Channel/ChannelPanel.tsx:20]:
```suggestion
    setInputText: (input: string) => void;
```

File: src/components/Compose/Channel/utils.ts (hunk 11 above)
Comment (nuanyang233) [reply src/components/Compose/Channel/utils.ts:59]: Please treat it as a separate hooks file.

File: src/components/Compose/Channel/utils.ts [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    imageUrl: '',
+    source: SocialPlatform.Farcaster,
+    url: '',
+    parentUrl: '',
+    followerCount: 0,
+    timestamp: 0,
+};
+export function isHomeChannel(channel: Channel) {
+    return channel.id === homeChannel.id;
+}
```
Comment (nuanyang233) [inline src/components/Compose/Channel/utils.ts:23]: It seems that we don't need `homeChannel`, when posting, the parameter about channel can be left empty.

File: src/services/postToFarcaster.ts (hunk 8 above)
Comment (nuanyang233) [reply src/services/postToFarcaster.ts:57]: Please don't do this. When posting a red envelope post, you must add the  redpacket channel. Unless the designer says to modify it.

File: src/store/useComposeStore.ts [hunk 14]
```diff
@@ -132,6 +136,7 @@ function createInitSinglePostState(cursor: Cursor): CompositePost {
         openGraphs: EMPTY_LIST,
         video: null,
         rpPayload: null,
+        channelMap: {} as Record<SocialPlatform, Channel | null>,
```
Comment (guanbinrui) [inline src/store/useComposeStore.ts:139]:
```suggestion
        channelMap: {
            [SocialPlatform.Farcaster]: null,
            [SocialPlatform.Lens]: null,
            [SocialPlatform.Twitter]: null,
        },
```

File: src/components/Compose/ChannelAction.tsx [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    channelList,
+    channel,
+    inputText,
+    setInputText,
+    updateChannel,
+}: ChannelActionProps) {
+    return (
+        <div className=" flex h-9 items-center justify-between pb-safe">
+            <span className=" text-[15px] text-secondary">
+                <Trans>Farcaster channel</Trans>
```
Comment (guanbinrui) [inline src/components/Compose/ChannelAction.tsx:27]:
```suggestion
                <Trans>${resolveSourceName(source)} channel</Trans>
```

File: src/helpers/channel.ts [hunk 16]
```diff
@@ -0,0 +1,27 @@
+import { HOME_CHANNEL } from '@/constants/channel.js';
+import { SocialPlatform } from '@/constants/enum.js';
+import { resolveSocialMediaProvider } from '@/helpers/resolveSocialMediaProvider.js';
+import type { Channel } from '@/providers/types/SocialMedia.js';
+
+export function isHomeChannel(channel: Channel) {
+    return channel.id === HOME_CHANNEL.id;
+}
+const channelCache: Record<string, Channel> = {};
```
Comment (guanbinrui) [inline src/helpers/channel.ts:9]: Use react-query.

File: src/hooks/useSearchChannel.ts [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    const [{ value }, searchChannel] = useAsyncFn(async () => {
+        return resolveSocialMediaProvider(source).searchChannels(kw);
+    }, [kw, source]);
+
+    useDebounce(async () => searchChannel(), 300, [searchChannel]);
+    return kw ? value?.data.filter(Boolean).slice(0, CHANNEL_SEARCH_LIST_SIZE) || ([] as Channel[]) : defaultChannels;
+}
+
+export function useDefaultChannelList(source: SocialPlatform) {
+    const { value: fireflyChannel } = useAsync(async () => fetchFireflyGardenChannel(source), []);
```
Comment (guanbinrui) [inline src/hooks/useSearchChannel.ts:21]: Dead-code fireflyChannel.

File: src/hooks/useSearchChannel.ts [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    const followedChannels = useFollowedChannels(source, CHANNEL_SEARCH_LIST_SIZE);
+    const { rpPayload } = useCompositePost();
+
+    switch (source) {
+        case SocialPlatform.Farcaster:
+            return [HOME_CHANNEL, rpPayload ? fireflyChannel : null, ...followedChannels]
+                .filter(Boolean)
```
Comment (guanbinrui) [inline src/hooks/useSearchChannel.ts:28]: Use `compact` from `lodash`.

File: src/hooks/useSearchChannel.ts [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+export function useDefaultChannelList(source: SocialPlatform) {
+    const { value: fireflyChannel } = useAsync(async () => fetchFireflyGardenChannel(source), []);
+    const followedChannels = useFollowedChannels(source, CHANNEL_SEARCH_LIST_SIZE);
+    const { rpPayload } = useCompositePost();
+
+    switch (source) {
+        case SocialPlatform.Farcaster:
+            return [HOME_CHANNEL, rpPayload ? fireflyChannel : null, ...followedChannels]
```
Comment (guanbinrui) [inline src/hooks/useSearchChannel.ts:27]: Dedup

File: src/hooks/useSearchChannel.ts [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+import type { Channel } from '@/providers/types/SocialMedia.js';
+
+export function useSearchChannels(kw: string, source: SocialPlatform) {
+    const defaultChannels = useDefaultChannelList(source);
+    const [{ value }, searchChannel] = useAsyncFn(async () => {
+        return resolveSocialMediaProvider(source).searchChannels(kw);
+    }, [kw, source]);
+
+    useDebounce(async () => searchChannel(), 300, [searchChannel]);
+    return kw ? value?.data.filter(Boolean).slice(0, CHANNEL_SEARCH_LIST_SIZE) || ([] as Channel[]) : defaultChannels;
```
Comment (guanbinrui) [inline src/hooks/useSearchChannel.ts:17]: compact from `lodash`.

**Comment by tomwang-io** [comment]
1.不同主题下输入框背景色 2.i18n中文显示还有点问题

File: src/components/Compose/ComposeAction.tsx [hunk 21]
```diff
@@ -92,6 +96,12 @@ export function ComposeAction(props: ComposeActionProps) {
     const maxImageCount = getCurrentPostImageLimits(availableSources);
     const mediaDisabled = !!video || images.length >= maxImageCount;
 
+    // channel
+    const currentSourceWithChannelSupport = SocialPlatform.Farcaster;
```
Comment (guanbinrui) [inline src/components/Compose/ComposeAction.tsx:100]: Move to constants.

File: src/components/Compose/ChannelSearchPanel.tsx [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+    inputText,
+    setInputText,
+    isLoading,
+    isError,
+}: ChannelSearchPanelProps) {
+    const isSmall = useIsSmall('max');
+
+    const listBox = isLoading ? (
+        <div className="m-auto">
+            <Trans>Loading...</Trans>
```
Comment (guanbinrui) [inline src/components/Compose/ChannelSearchPanel.tsx:37]: Use `<Loading />`.

File: src/constants/channel.ts [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+    imageUrl: '',
+    source: SocialPlatform.Farcaster,
+    url: '',
+    parentUrl: '',
+    followerCount: 0,
+    timestamp: 0,
+};
+
+export const CHANNEL_SEARCH_LIST_SIZE = 10;
+export const FF_GARDEN_CHANNEL_ID = 'firefly-garden';
```
Comment (guanbinrui) [inline src/constants/channel.ts:16]:
```suggestion
export const FF_GARDEN_CHANNEL = {
    name: 'Firefly Garden',
    id: 'firefly-garden',
    ...
};
```

File: src/hooks/useFetchChannel.ts [hunk 24]
```diff
@@ -0,0 +1,14 @@
+import { useQuery } from '@tanstack/react-query';
+
+import type { SocialPlatform } from '@/constants/enum.js';
+import { resolveSocialMediaProvider } from '@/helpers/resolveSocialMediaProvider.js';
+
+export function useFetchChannel(chanelId: string, source: SocialPlatform) {
```
Comment (guanbinrui) [inline src/hooks/useFetchChannel.ts:6]: Not necessary. Hooks are costy.

File: src/hooks/useSearchChannel.ts [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+ */
+export function useFollowedChannels(source: SocialPlatform, count: number) {
+    const { data } = useQuery({
+        queryKey: ['searchFollowChannel', source],
+        queryFn: () => {
+            return resolveSocialMediaProvider(source).searchChannels('');
+        },
+    });
+    return data?.data.slice(0, count) || ([] as Channel[]);
+}
```
Comment (guanbinrui) [inline src/hooks/useSearchChannel.ts:61]: Over modularity, merge those hooks into `useSearchChannels`.

File: src/services/postToFarcaster.ts [hunk 26]
```diff
@@ -52,8 +54,8 @@ export async function postToFarcaster(type: ComposeType, compositePost: Composit
                 (x) => x.url.toLowerCase(),
             ),
             commentOn: type === 'reply' && farcasterParentPost ? farcasterParentPost : undefined,
-            parentChannelKey: hasPayload ? env.external.NEXT_PUBLIC_REDPACKET_CHANNEL_KEY : undefined,
-            parentChannelUrl: hasPayload ? env.external.NEXT_PUBLIC_REDPACKET_CHANNEL_URL : undefined,
+            parentChannelKey: channel ? (isHomeChannel(channel) ? undefined : channel.id) : undefined,
```
Comment (guanbinrui) [inline src/services/postToFarcaster.ts:57]: If `NEXT_PUBLIC_REDPACKET_CHANNEL_KEY` and `NEXT_PUBLIC_REDPACKET_CHANNEL_URL` are not used anymore, please remove them from  `env.ts`.

File: src/hooks/useSearchChannel.ts [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+    const { data, isLoading, isError } = useQuery({
+        enabled: !!debouncedKw,
+        queryKey: ['searchChannels', source, debouncedKw],
+        queryFn: () => resolveSocialMediaProvider(source).searchChannels(debouncedKw),
+    });
+
+    if (!debouncedKw) return { channelList: defaultChannels, isLoading: false, isError: false };
+    if (!data) return { channelList: EMPTY_LIST, isLoading, isError };
+
+    return { channelList: compact(data.data).slice(0, CHANNEL_SEARCH_LIST_SIZE), isLoading, isError: false };
```
Comment (guanbinrui) [inline src/hooks/useSearchChannel.ts:31]: Don't rename queryResult. Just use queryResult directly.

File: src/components/Compose/ComposeAction.tsx [hunk 28]
```diff
@@ -248,6 +258,21 @@ export function ComposeAction(props: ComposeActionProps) {
                     )}
                 </Popover>
             </div>
+            {showChannel && channel ? (
+                <ChannelAction
+                    source={currentSourceWithChannelSupport}
+                    isRootPost={isRootPost}
+                    channelList={channelList}
+                    isError={isError}
+                    isLoading={isLoading}
```
Comment (guanbinrui) [inline src/components/Compose/ComposeAction.tsx:267]: Don't pass down query state variables. It's error-prone. Just do querying in `<ChannelAction />`.

File: src/store/useComposeStore.ts [hunk 29]
```diff
@@ -132,6 +137,11 @@ function createInitSinglePostState(cursor: Cursor): CompositePost {
         openGraphs: EMPTY_LIST,
         video: null,
         rpPayload: null,
+        channelMap: {
```
Comment (guanbinrui) [inline src/store/useComposeStore.ts:140]: Align naming:

```suggestion
        channel: {
```

File: src/components/Compose/ChannelSearchPanel.tsx [hunk 30]
```diff
@@ -0,0 +1,96 @@
...
+                    }}
+                >
+                    <div className={'flex h-[24px] items-center gap-2'}>
+                        <Avatar
+                            className="mr-3 shrink-0 rounded-full border"
+                            src={channel.imageUrl}
+                            size={isSmall ? 24 : 24}
+                            alt={channel.name}
+                        />
+                        <span className={classNames(' font-bold text-main')}>{channel.name}</span>
```
Comment (guanbinrui) [inline src/components/Compose/ChannelSearchPanel.tsx:61]:
```suggestion
                        <span className=" font-bold text-main">{channel.name}</span>
```

File: src/components/Compose/ChannelSearchPanel.tsx [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+            <Fragment key={channel.id}>
+                <div
+                    className={'flex h-[32px] cursor-pointer items-center justify-between'}
+                    onClick={() => {
+                        if (channel.id !== selectedChannel?.id) {
+                            selectChannel(channel);
+                        }
+                    }}
+                >
+                    <div className={'flex h-[24px] items-center gap-2'}>
```
Comment (guanbinrui) [inline src/components/Compose/ChannelSearchPanel.tsx:54]:
```suggestion
                    <div className='flex h-[24px] items-center gap-2'>
```

File: src/components/Compose/ChannelSearchPanel.tsx [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+                        if (ch.id !== channel?.id) {
+                            updateChannel(source, ch);
+                        }
+                    }}
+                >
+                    <div className="flex h-[24px] items-center gap-2">
+                        <Avatar
+                            className="mr-3 shrink-0 rounded-full border"
+                            src={ch.imageUrl}
+                            size={isSmall ? 24 : 24}
```
Comment (guanbinrui) [inline src/components/Compose/ChannelSearchPanel.tsx:54]:
```suggestion
                            size={24}
```

**Pull request merged**